docs/ROADMAP.md

```diff
@@ -0,0 +1,382 @@
+VISION describes, provides **Correct** connection management as it pertains to
+HTTP.
+
+There will be still be a type to wrap the different versions. It will no longer
+be generic over the transport type, to prepare for being able to wrap HTTP/3
+connections. Exactly how it will wrap, either by using internal trait objects,
+or an `enum Either` style, or using a `trait Connection` that each type
+implements, is something to be determined. It's likely that this "auto" type
+will start in `hyper-util`.
+
+### Focus on the `Connection` level
+
+As mentioned in the *Known Issues*, the higher-level `Client` and `Server` have
+stability and complexity problems. Therefore, for hyper 1.0, the main API will
+focus on the "lower-level" connection types. The `Client` and `Server` helpers
+will be moved to `hyper-util`.
+
+## Public API
+
+### body
+
+The `Body` struct is removed. Its internal "variants" are [separated into
+distinct types](https://github.com/hyperium/hyper/issues/2345), and can start
+in either `hyper-util` or `http-body-util`.
+
+The exported trait `HttpBody` is renamed to `Body`.
+
+A single `Body` implementation in `hyper` is the one provided by receiving
+client responses and server requests. It has the name `Streaming`.
+
+> **Unresolved**: Other names can be considered during implementation. Another
+> option is to not publicly name the implementation, but return `Response<impl
+Body>`s.
+
+The `Body` trait will be experimented on to see about making it possible to
+return more frame types beyonds just data and trailers.
+
+> **Unresolved**: What exactly this looks like will only be known after
+> experimentation.
+
+### client
+
+The high-level `hyper::Client` will be removed, along with the
+`hyper::client::connect` module. They will be explored more in `hyper-util`.
+
+As described in *Design*, the `client::conn` module will gain `http1` and
+`http2` sub-modules, providing per-version `SendRequest`, `Connection`, and
+`Builder` structs. An `auto` version can be explored in `hyper-util`.
+
+### error
+
+The `hyper::Error` struct remains in place.
+
+All errors returned from `Error::source()` are made opaque. They are wrapped an
+internal `Opaque` newtype that still allows printing, but prevents downcasting
+to the internal dependency.
+
+A new `hyper::error::Code` struct is defined. It is an opaque struct, with
+associated constants defining various code variants.
+
+> Alternative: define a non-exhaustive enum. It's not clear that this is
+> definitely better, though. Keeping it an opaque struct means we can add
+> secondary parts to the code in the future, or add bit flags, or similar
+> extensions.
+
+The purpose of `Code` is to provide an abstraction over the kind of error that
+is encountered. The `Code` could be some behavior noticed inside hyper, such as
+an incomplete HTTP message. Or it can be "translated" from the underlying
+protocol, if it defines protocol level errors. For example, an
+`h2::Reason::CANCEL`.
```

Comment:
My concern with any division between errors we use internally and errors that we expose in the public API is how far the abstraction gap is between the two. I think we agree that there's currently too large of a gap between the error details and the `is_<kind>()` predicates that are the primary way to inspect them. What strategy could we use to prevent `Code` from ending up in a similar position? 

Perhaps `Code` could be part of the internal error definition rather than an abstraction around it, while still leaving us with enough internal fields for implementation flexibility? Like if `hyper::error::Kind` was part of the public API. I suppose this basically ends up being the non-exhaustive enum alternative, though...

Reply:
This is an interesting topic I'd like to hear more about.

I think keeping the two types separate (internal `Kind` and public `Code`) provides flexibility. The current `Kind` enum uses sub-enums to provide more details. So far we've used new variants when the message should be just a little different, but the general error is still the same. Instead of storing a `&'static str` inside the `Error`, we just use different variants to allow printing the different string. I don't believe programmatically know those minor variant differences is useful to a user.

Personally, I subscribe to the thought that there are a very small class of errors that a user can programmatically do something about, and the rest are only addressable by a human looking at the output message. Thus, I'd want the `Code` to be able to represent those, but otherwise not give more detail than needed since it can make code fragile to internal changes.

Reply:
>Personally, I subscribe to the thought that there are a very small class of errors that a user can programmatically do something about, and the rest are only addressable by a human looking at the output message.

I agree when thinking about "what do we do with this request", but our primary desire for programmatic access is to increase the fidelity of metrics that we report to our customers. Our proxy and even the customer's code they might not decide to do anything differently during the course of one request based on an error. But we do have to report that error back to the customer in a format that doesn't assume it's only for human consumption.

For example, some of our customers depend on being able to distinguish types of errors in downstream automated systems that make decisions about, e.g., how much % of their traffic should go through a particular service at a given time. There are certain classes of errors and variations on those classes that they're particularly sensitive to; just one invalid HTTP header, if it comes from their origin, might be enough to warrant shifting traffic, whereas a certain % of generic connection failures might be acceptable.

It might be that handling these on a case-by-case basis and being eagerly willing to extend `Code` is the best approach, and we can try to start `Code` off with enough variants to cover all of the logic we're currently aware of customers using.